parser_utils.py

```diff
@@ -0,0 +1,331 @@
+from enum import Enum
+
+class Lists(Enum):
+    """Enum class for representing various list types that appear in the grammar."""
+    GOPLIST = 0
+    ANYLIST = 1
+    IDLIST = 2
+    MIXEDLIST = 3
+    EXPLIST = 4
+
+
+class Ops(Enum):
+    """Enum class for representing various Op types that appear in the grammar."""
+    GATE = 0
+    OPERATOR = 1
+    CHANNEL = 2
+    MEASURE = 3
+    BARRIER = 4
+
+
+class Declarations(Enum):
+    """Enum class for representing various declarations that are allowed by the grammar."""
+    QREG = 0
+    CREG = 1
+    GATE = 2
+    OPERATOR = 3
+    CHANNEL = 4
+
+
+class ParsedList:
+    """Bookkeeping class for the various list-like objects encountered while parsing."""
+    def __init__(self, name, list_):
+        self.name = name
+        self.list = list_
```

Comment:
Question: why is this needed? I don't quite see it...

Reply:
There are so many situations in the grammar that depend on the specific type of "list" you are getting. This class allows us to keep that basic information about the list type

Reply:
You said ":pray: :100:" above where this was being used :laughing:

Reply:
I didn't realize this was what the implementation looked like until I scrolled further down 😆